src/class-cache-collector.php

```diff
@@ -7,9 +7,517 @@
+
+	/**
+	 * Handle a post update and purge the cache if the post being updated is
+	 * newer than the threshold.
+	 *
+	 * @param int $post_id Post ID.
+	 */
+	public static function on_post_update( int $post_id ) {
+		$post = get_post( $post_id );
+
+		if ( ! $post ) {
+			return;
+		}
+
+		$threshold = static::$post_update_threshold;
+
+		/**
+		 * Filter the threshold for cache key expiration.
+		 *
+		 * @param int $threshold Threshold in seconds.
+		 */
+		$threshold = (int) apply_filters( 'cache_collector_post_threshold', $threshold, $post_id );
+
+		// If the post is newer than the threshold, purge the cache.
+		if ( get_the_date( 'U', $post ) > ( time() - $threshold ) ) {
+			static::for_post( $post )->purge();
+		}
+	}
+
+	/**
+	 * Register the post type for the cache collector.
+	 */
+	public static function register_post_type() {
+		register_post_type( // phpcs:ignore WordPress.NamingConventions.ValidPostTypeSlug.NotStringLiteral
+			static::POST_TYPE,
+			[
+				'public'             => false,
+				'publicly_queryable' => false,
+			]
+		);
+	}
+
+	/**
+	 * Cleanup the cache collector post stored in the database.
+	 *
+	 * Checks if the post is older than the threshold and if so, deletes it. If
+	 * it is not older than the threshold, it will check if the keys in the
+	 * collection it is storing are expired.
+	 */
+	public static function cleanup() {
+		$page  = 1;
+		$limit = 100;
+
+		while ( true ) {
+			if ( $page > $limit ) {
+				if ( function_exists( 'ai_logger' ) ) {
+					ai_logger()->warning( 'Cache Collector: Reached limit of posts to check.' );
+				}
+
+				break;
+			}
+
+			$posts = get_posts( // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.get_posts_get_posts
+				[
+					'date_query'       => [
+						[
+							'column' => 'post_modified',
+							'before' => gmdate( 'Y-m-d H:i:s', time() - static::$expiration_threshold ),
+						],
+					],
+					'paged'            => $page++,
+					'post_type'        => static::POST_TYPE,
+					'posts_per_page'   => 100,
+					'suppress_filters' => false,
+				]
+			);
+
+			if ( empty( $posts ) ) {
+				break;
+			}
+
+			foreach ( $posts as $post ) {
+				$collection = get_post_meta( $post->ID, '_collection', true );
+
+				if ( empty( $collection ) ) {
+					wp_delete_post( $post->ID, true );
+					continue;
+				}
+
+				( new static( $collection, $post ) )->save();
+			}
+		}
+	}
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string               $collection Cache collection to attach to.
+	 * @param WP_Post|WP_Term|null $parent Object to attach to, optional. Used for storage of the cache collection.
+	 * @param LoggerInterface|null $logger Logger to use.
+	 */
+	public function __construct(
+		public string $collection,
+		public WP_Post|WP_Term|null $parent = null,
+		public ?LoggerInterface $logger = null,
+	) {
+	}
+
+	/**
+	 * Save the collector on destruct.
+	 */
+	public function __destruct() {
+		if ( ! empty( $this->pending_keys ) ) {
+			$this->save();
```

Comment:
If I'm reading correctly, this logic implies that any calls to register a cache key for a given object will result in a database query during the same request, since `__destruct()` will always run, and since the logic in `save()` sets (or bumps) the expiration time for pending keys, thus creating an updated value to save to post or term meta.

If so, that's a potential gotcha that might be worth documenting: Cache keys shouldn't be "always on" — they should be registered only when needed, not just for the reason above but because the expiration time will get bumped each time, too.